bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -528,6 +547,45 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
 	return ctrl.Result{}, nil
 }
 
+// resolveFiles maps .Spec.Files into cloudinit.Files, resolving any object references
+// along the way.
+func (r *KubeadmConfigReconciler) resolveFiles(ctx context.Context, cfg *bootstrapv1.KubeadmConfig, merge ...bootstrapv1.File) ([]bootstrapv1.File, error) {
+	collected := append(cfg.Spec.Files, merge...)
+
+	for i := range collected {
+		in := collected[i]
+		if in.ContentFrom != nil {
+			data, err := r.resolveSecretFileContent(ctx, cfg.Namespace, in)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to resolve file source")
+			}
+			in.ContentFrom = nil
+			in.Content = string(data)
+			collected[i] = in
```

Comment:
Just to double check, this logic is only changing the content of the struct at _runtime_, but it won't be stored in the KubeadmConfig at the end when we patch.

Reply:
That's the intent. I could probably make this a little bit more explicit by newing up a slice, copying, and returning? Plus comments